### ISSUE TITLE
navigation bindings in menus

### ISSUE THREAD

**Issue by ds2606** [KoffeinFlummi/htop-vim#5]
Would it be possible to integrate vim navigation bindings in the htop menus, e.g. choosing a sort order?

**Comment by KoffeinFlummi** [comment]
I considered it when I first forked this ages ago, but since I only use those menus very occasionally, I figured it wasn't worth the effort. I don't have much time atm, but I would gladly accept a PR.

**Comment by z0noxz** [comment]
I created a PR #6 for the settings menu. I haven't yet had time to check the menu for sort order.

**Comment by z0noxz** [comment]
I've also added a similar PR #7 for the sort order menu now.

**Comment by ds2606** [comment]
Unfortunately, the PR broke functionality in the search: either incremental name search or incremental name filtering (searches beginning with `/` or `\`) don't accept the `hjkl` keys. Try it out: search for any process name with an `h` `j` `k` or `l` and the search quits out. Can't find a way to escape those characters in - I'm pulling and recompiling an older commit for now.

Thank you for helping on this though!

**Comment by z0noxz** [comment]
I see. I guess a state check for `search mode` before the `switch` would fix the issue. I'll have a look later this weekend.

**Comment by z0noxz** [comment]
This seems to fix the issue (in ScreenManager.c):
```
   if (!result) <------------
   switch (ch) {
      case 'h': case KEY_ALT('H'): ch = KEY_LEFT; break;
      case 'j': case KEY_ALT('J'): ch = KEY_DOWN; break;
      case 'k': case KEY_ALT('K'): ch = KEY_UP; break;
      case 'l': case KEY_ALT('L'): ch = KEY_RIGHT; break;
    }
```

Haven't tested everything, so I won't add a PR just yet...
`

**Comment by z0noxz** [comment]
I was way to quick. This is NOT a fix. I'll have a fresh look at it this weekend.

**Comment by z0noxz** [comment]
Is there any indication for which mode is active such as Search or Filter, like some global status indicator?

On way I have managed to solve it is using something like:

```
FunctionBar* bar = panelFocus->currentBar;
char *fnc = bar->functions[bar->size - 1];
if (!(strncmp(fnc, " Search:", 8) == 0 || strncmp(fnc, " Filter:", 8) == 0)) {
...
```
However this is not particularly clean, and can be prone to bugs

EDIT:
Could a solution be to add such a mode indicator in Settings_ struct?

**Comment by ds2606** [comment]
I think I might've found a fix, in the conditional at line 218. I moved the switch inside the check for the `panelFocus`, and added a conditional to check the current pane type, as follows:
```c
      if (Panel_eventHandlerFn(panelFocus)) {
         if (Panel_eventHandlerFn(panelFocus) != MainPanel_eventHandler) {
            switch (ch) {
               case 'h': case KEY_ALT('H'): ch = KEY_LEFT; break;
               case 'j': case KEY_ALT('J'): ch = KEY_DOWN; break;
               case 'k': case KEY_ALT('K'): ch = KEY_UP; break;
               case 'l': case KEY_ALT('L'): ch = KEY_RIGHT; break;
            }
         }
         result = Panel_eventHandler(panelFocus, ch);
      }
```
It seems that inc and normal search happen in the main panel, so for all other panels (i.e. where settings/sort menus happen, which is what the original PR is for) it's fine to translate an `hjkl` char input to a key movement in the switch statement. The only problem is, I'm not sure the check will work, since I don't know what the type of `MainPanel_eventHandler` is. The only reference to the type definition is line 42 of `Panel.h`, where a `PanelClass_` struct has a `const Panel_EventHandler eventHandler`. I can't find the typedef of `Panel_EventHandler` anywhere in the code base. Does anyone know what exactly a `Panel_EventHandler` is? It looks like a function pointer maybe? If this boolean comparison above works though the whole thing should run I think.

Edit: yep, it's a function pointer, defined in line 35 of `Panel.h`:
```c
typedef HandlerResult(*Panel_EventHandler)(Panel*, int);
```
Can we do boolean compares with function pointers?

Edit2: yes we can, this ended up working out very nicely, using your idea z0noxz of trying to parse out which mode we're in at a given moment. I ended up making the address of the `MainPanel_eventHandler` function visible from `MainPanel.c` and `MainPanel.h` to make this check happen, but that's the only significant change that had to happen to fix this. Again, I'm new to git/github so it's probably worth checking that nothing here is adverse in the long run!

Edit3: created PR [#9](https://github.com/KoffeinFlummi/htop-vim/pull/9)
